### ISSUE TITLE
Keyrefs to content elements inside `@copy-to` topics result in bad HTML links

### ISSUE THREAD

**Issue by chrispy-snps** [dita-ot/dita-ot#4076]
## Expected Behavior
Started from:
https://www.oxygenxml.com/forum/post68679.html#p68679

If I have a `@copy-to` version of a topic:

```
<topicref href="topic.dita" keys="topic1"/>
<topicref href="topic.dita" keys="topic2" copy-to="topic2.dita"/>
```

and I have keyrefs to content inside that topic:

```
<p><xref keyref="topic1/my-fig-id"/></p>
<p><xref keyref="topic2/my-fig-id"/></p>
```

then the references should resolve properly.

## Actual Behavior

References to content elements inside the `@copy-to` element have a `null` component in their anchor names, causing the link to be invalid:

```
<p class="p"><a class="xref" href="topic.html#common__my-fig-id">Figure 1</a></p>
<p class="p"><a class="xref" href="topic2.html#null__my-fig-id">topic2.html#null__my-fig-id</a></p>
```

References to the topic work correctly; only references to content inside the topic are bad.

## Possible Solution
Not sure yet.

## Steps to Reproduce
1. Download the following testcase:
   [4076.zip](https://github.com/dita-ot/dita-ot/files/10200364/4076.zip)
2. Run the following command:
   `dita --project.xml`

## Environment
* DITA-OT version: 4.0
* Operating system and version: Linux
* How did you run DITA-OT? `dita` command
* Transformation type: HTML5

**Comment by chrispy-snps** [comment]
It looks like the issue is in `KeyrefModule`, somehow the `@id` is not resolved for the reference into the `@copy-to` topic:

![4076](https://user-images.githubusercontent.com/50950969/206862100-2c5639ce-9b84-471d-9254-5b48c242042b.svg)

**Comment by raducoravu** [comment]
When publishing the keyref stage takes place before the copy-to stage. So the file "topic2.dita" does not yet exist in the temporary files folder.
In the method org.dita.dost.writer.KeyrefPaser.processElement(Attributes) there is a call to:

      topicId = getFirstTopicId(topicFile);

which is called for the href field of the key definition:

      final URI href = keyDef.href;

Looking at how keys are created "KeyrefReader.readKeyDefinition(XdmNode, Map<String, KeyDef>)" the "href" field actually points to the @copy-to attribute value which does not point to a topic that already exists on disk.
So maybe on the "KeyrefPaser.processElement(Attributes)" method this fallback code could be added if we cannot locate the topicId:

                                    topicId = getFirstTopicId(topicFile);
                                    if(topicId == null) {
                                        String directHref = keyDef.element.attribute(DITAConstants.HREF_ATTR_NAME);
                                        if(directHref != null && ! directHref.equals(href.toString())) {
                                            //Try to also resolve the topic ID using the direct href
                                            try {
                                                URI directTarget = keyDef.source.resolve(new URI(directHref));
                                                URI directTopicFile = currentFile.resolve(stripFragment(directTarget));  
                                                topicId = getFirstTopicId(directTopicFile);
                                            } catch (URISyntaxException e) {
                                                //Ignore
                                            }
                                        }
                                    }

**Comment by chrispy-snps** [comment]
@raducoravu - I had to change the code a bit:

```
$ git diff src/main/java/org/dita/dost/writer/KeyrefPaser.java
diff --git a/src/main/java/org/dita/dost/writer/KeyrefPaser.java b/src/main/java/org/dita/dost/writer/KeyrefPaser.java
index 92c55a32d..378c947aa 100644
--- a/src/main/java/org/dita/dost/writer/KeyrefPaser.java
+++ b/src/main/java/org/dita/dost/writer/KeyrefPaser.java
@@ -23,6 +23,7 @@ import org.xml.sax.helpers.AttributesImpl;

 import java.io.File;
 import java.net.URI;
+import java.net.URISyntaxException;
 import java.util.*;

 import static java.util.Arrays.asList;
@@ -535,6 +536,19 @@ public final class KeyrefPaser extends AbstractXMLFilter {
                                 String topicId = null;
                                 if (relativeTarget.getFragment() == null && !"".equals(elementId)) {
                                     topicId = getFirstTopicId(topicFile);
+                                    if(topicId == null) {
+                                        String directHref = keyDef.element.attribute(ATTRIBUTE_NAME_HREF);
+                                        if(directHref != null && ! directHref.equals(href.toString())) {
+                                            //Try to also resolve the topic ID using the direct href
+                                            try {
+                                                URI directTarget = keyDef.source.resolve(new URI(directHref));
+                                                URI directTopicFile = currentFile.resolve(stripFragment(directTarget));
+                                                topicId = getFirstTopicId(directTopicFile);
+                                            } catch (URISyntaxException e) {
+                                                //Ignore
+                                            }
+                                        }
+                                    }
                                 }
                                 final URI targetOutput = normalizeHrefValue(relativeTarget, elementId, topicId);
                                 XMLUtils.addOrSetAttribute(resAtts, refAttr, targetOutput.toString());
```

and with your fix, the testcase publishes successfully:

```
<p class="p">Figure in original topic: <a class="xref" href="topic.html#common__my-fig-id">Figure 1</a></p>
<p class="p">Figure in copy-to topic: <a class="xref" href="topic2.html#common__my-fig-id">Figure 1</a></p>
```

**Comment by raducoravu** [comment]
Thanks for testing this @chrispy-snps I will try to see when I have the time to look into how building an automatic test for this would look like.

**Comment by infotexture** [comment]
Fixed in [4.0.2](https://github.com/dita-ot/dita-ot/releases/tag/4.0.2) with https://github.com/dita-ot/dita-ot/pull/4110.
